cpp/ql/src/semmle/code/cpp/models/interfaces/FormattingFunction.qll

```diff
@@ -115,7 +115,19 @@ abstract class FormattingFunction extends Function {
    * Gets the position of the first format argument, corresponding with
    * the first format specifier in the format string.
    */
-  int getFirstFormatArgumentIndex() { result = getNumberOfParameters() }
+  int getFirstFormatArgumentIndex() {
+    result = getNumberOfParameters() and
+    // the formatting function either has a definition in the snapshot, or all
+    // `DeclarationEntry`s agree on the number of parameters (otherwise we don't
+    // really know the correct number)
+    (
+      hasDefinition()
+      or
+      forall(FunctionDeclarationEntry fde | fde = getADeclarationEntry() |
+        result = fde.getNumberOfParameters()
+      )
```

Comment:
Why not put this check in the charpred of `FormattingFunction` instead? Then we're sure that any use of it will be impossible when it's ambiguous, including use by other queries than "wrong number of arguments". Also, you wouldn't have to add a check to `FormattingFunctionCall` since there can be no calls to a `FormattingFunction` that doesn't exist.

`NonConstantFormat.ql` uses `FormattingFunction` because it needs to find out which argument is the format string, but it never checks which argument is the first vararg. See https://github.com/Semmle/ql/blob/b31cd8ab3233c14ee5ae404952e710c1e61d5120/cpp/ql/src/Likely%20Bugs/Format/NonConstantFormat.ql#L126-L128

Reply:
WrongTypeFormatArguments.ql is also affected and I've [just] written a separate fix for it.  Next week I'll try adding the tests from that to this PR, and see if updating the charpred as you suggest fixes both problems at once.

Reply:
> Why not put this check in the charpred of `FormattingFunction` instead?

I'm going off this approach now.  `FormattingFunction` is an abstract class so I have to invent a parent class above it to insert the characteristic predicate.  Also denying a check that something is `instanceof FormattingFunction` just because we can't figure out the parameters feels wrong.

> `NonConstantFormat.ql` uses `FormattingFunction` because it needs to find out which argument is the format string, but it never checks which argument is the first vararg.

So in this case, we'd potentially lose good results by constraining at the characteristic predicate instead of `getNumFormatArgument()`.

Reply:
If it's ambiguous which parameter is the format string, then `NonConstantFormat.ql` would also get false positives from non-constant values that reach a parameter that isn't the format string.

Reply:
I think my follow up PR for WrongTypeFormatArguments.ql will address that.  I should add a test case to confirm correct behaviour on NonConstantFormat.ql to that PR also.